src/hypothesis/strategies.py

```diff
@@ -982,32 +987,61 @@ def fractions(min_value=None, max_value=None, max_denominator=None):
     of any generated values is no greater than max_denominator. Note that
-    max_denominator must be at least 1.
+    max_denominator must be None or a positive integer, and the denominator
+    of min_value and max_value may not be greater than max_denominator.
 
     """
+    if min_value is not None:
+        min_value = Fraction(min_value)
+    if max_value is not None:
+        max_value = Fraction(max_value)
+
     check_valid_bound(min_value, 'min_value')
     check_valid_bound(max_value, 'max_value')
     check_valid_interval(min_value, max_value, 'min_value', 'max_value')
 
-    check_valid_integer(max_denominator)
-    if max_denominator is not None and max_denominator < 1:
-        raise InvalidArgument(
-            u'Invalid denominator bound %s' % max_denominator
-        )
-
-    denominator_strategy = integers(min_value=1, max_value=max_denominator)
-
     def dm_func(denom):
-        max_num = max_value * denom if max_value is not None else None
-        min_num = min_value * denom if min_value is not None else None
+        """Take denom, construct numerator strategy, and build fraction."""
+        # Four cases of algebra to get integer bounds and scale factor.
+        min_num, max_num = None, None
+        if max_value is None and min_value is None:
+            pass
+        elif min_value is None:
+            max_num = denom * max_value.numerator
+            denom *= max_value.denominator
+        elif max_value is None:
+            min_num = denom * min_value.numerator
+            denom *= min_value.denominator
+        else:
+            min_num = min_value.numerator * denom * max_value.denominator
+            max_num = max_value.numerator * denom * min_value.denominator
+            denom *= min_value.denominator * max_value.denominator
 
         return builds(
             Fraction,
             integers(min_value=min_num, max_value=max_num),
             just(denom)
         )
 
-    return denominator_strategy.flatmap(dm_func)
+    if max_denominator is None:
+        if min_value is not None and min_value == max_value:
+            return just(min_value)
+        return integers(min_value=1).flatmap(dm_func)
+
+    check_valid_integer(max_denominator)
+    if max_denominator < 1:
+        raise InvalidArgument('Denominator=%r must be >= 1' % max_denominator)
+    if min_value is not None and min_value.denominator > max_denominator:
+        raise InvalidArgument('min_value=%r incompatible with max_denominator'
+                              '=%r' % (min_value, max_denominator))
+    if max_value is not None and max_value.denominator > max_denominator:
+        raise InvalidArgument('max_value=%r incompatible with max_denominator'
+                              '=%r' % (max_value, max_denominator))
+
+    if min_value is not None and min_value == max_value:
+        return just(min_value)
+    return integers(1, max_denominator).flatmap(dm_func)\
```

Comment:
There were two intertwined problems:

1. The value bounds might not be integers.  If they're close together, integers() might be given an interval with no integers in it and therefore fail.  To correct this, we convert our bounds to fractions up-front and do a little more algebra in dm_func.

2. Now that we're doing the extra algebra, dm_func can actually increase our denominator - potentially past the max_denominator bound.  We therefore use the limit_denominator to clip it to the nearest fraction with at most that denominator, which is always within bounds by construction.